CMakeLists.txt

```diff
@@ -40,35 +40,44 @@ set( HEADER_FILES
 	${HEADER_FOLDER}/date/tz_private.h
 )
 
-if( BUILD_TZ_STATIC )
-	add_library( tz STATIC ${HEADER_FILES} ${SOURCE_FOLDER}/tz.cpp )
-else( )
-	add_library( tz SHARED ${HEADER_FILES} ${SOURCE_FOLDER}/tz.cpp )
-endif( )
-
 if( USE_SYSTEM_TZ_DB )
-	target_compile_definitions( tz PRIVATE -DUSE_AUTOLOAD=0 )
-	target_compile_definitions( tz PRIVATE -DHAS_REMOTE_API=0 )
+	set(AUTO_DOWNLOAD 0)
+	set(HAS_REMOTE_API 0)
+
 	# cannot set USE_OS_TZDB to 1 on Windows
 	if( NOT WIN32 )
-		target_compile_definitions( tz PUBLIC -DUSE_OS_TZDB=1 )
+		set(USE_OS_TZDB 1)
+	else ( )
+		set(USE_OS_TZDB 0)
 	endif( )
 else( )
-	target_compile_definitions( tz PRIVATE -DUSE_AUTOLOAD=1 )
-	target_compile_definitions( tz PRIVATE -DHAS_REMOTE_API=1 )
-	target_compile_definitions( tz PUBLIC -DUSE_OS_TZDB=0 )
+	set(AUTO_DOWNLOAD 1)
+	set(HAS_REMOTE_API 1)
+	set(USE_OS_TZDB 0)
+
 	find_package( CURL REQUIRED )
 	include_directories( SYSTEM ${CURL_INCLUDE_DIRS} )
 	set( OPTIONAL_LIBRARIES ${CURL_LIBRARIES} )
 endif( )
 
+if( NOT TZ_CXX_STANDARD )
+	set( TZ_CXX_STANDARD 17 )
+endif( )
+
+set(TZ_CONFIG_H_PATH ${CMAKE_CURRENT_BINARY_DIR}/${HEADER_FOLDER}/date/tz_config.h)
+configure_file(tz_config.h.in ${TZ_CONFIG_H_PATH} @ONLY)
+
+
+if( BUILD_TZ_STATIC )
```

Comment:
We can simplify this. See https://cmake.org/cmake/help/v3.0/command/add_library.html.

If `BUILD_TZ_STATIC` just sets `BUILD_SHARED_LIBS` to the correct value and type is not explicitly set in this call.

Reply:
Could definitely be a separate change though.

Reply:
I put that change in here for reference: https://github.com/HowardHinnant/date/pull/316

Reply:
One thing to consider is, if tz.cpp is e.g. compiled at C++11 standard, and then tz.hpp is used in a compilation unit that compiles at C++17 standard, is that going to cause ODR violations? Especially if e.g. `HAS_STRINGVIEW` is true in one version and false in the other?

It might be a good idea to make standard detection part of the configuration header, for tz lib, to avoid such problems. It's likely going to be more work than this PR was though, up to you guys if it's worth it.

Reply:
Currently a utility using tz.h with c++ without `std::string_view` will use/link functions with `std::string` arguments. That will fail if the library was built with `std::string_view` arguments unless there is some bridging malarkey in the string_view stuff that I do not know about (possible).

Perhaps the library should make symbols for `std::string` and `std::string_view` if possible? Then it could be linked against utilities using either (the switch would be in the header).

Using the default `CMAKE_CXX_STANDARD` does mean that anyone who ties this library into their build system (with `add_subdirectory`) will automatically build it with the same `CXX_STANDARD` as the outer utility.

Reply:
Yeah, I don't disagree that `CMAKE_CXX_STANDARD` should be used.

I'm looking through the code now...

I guess I'm worried about something like this:

https://github.com/HowardHinnant/date/blob/master/include/date/tz.h#L101

```
#ifndef HAS_DEDUCTION_GUIDES
#  if __cplusplus >= 201703
#    define HAS_DEDUCTION_GUIDES 1
#  else
#    define HAS_DEDUCTION_GUIDES 0
#  endif
#endif // HAS_DEDUCTION_GUIDES
```

I'd rather have `cmake` figure out this value, using `CMAKE_CXX_STANDARD` at the time that date is configured, and then write a variable `HAS_DEDUCTION_GUIDES` into `tz_config.h`, so that it has consistent configuration both when it is built and when it is consumed.